sway/input/text_input.c

```diff
@@ -255,6 +268,220 @@ static void relay_handle_text_input(struct wl_listener *listener,
+		parent.height = view->geometry.height;
+		popup->scene_tree = wlr_scene_subsurface_tree_create(root->layers.popup, popup->popup_surface->surface);
+		popup->desc.view = view;
+	}
+
+	struct wlr_scene_tree *relative = wlr_scene_tree_create(relative_parent);
+
+	popup->desc.relative = &relative->node;
+	if (!scene_descriptor_assign(&popup->scene_tree->node,
+			SWAY_SCENE_DESC_POPUP, &popup->desc)) {
+		wlr_scene_node_destroy(&popup->scene_tree->node);
+		popup->scene_tree = NULL;
+		return;
+	}
+
+	if (!cursor_rect) {
+		cursor_area.x = 0;
+		cursor_area.y = 0;
+		cursor_area.width = parent.width;
+		cursor_area.height = parent.height;
+	}
+
+	int popup_width = popup->popup_surface->surface->current.width;
+	int popup_height = popup->popup_surface->surface->current.height;
+	int x1 = parent.x + cursor_area.x;
+	int x2 = parent.x + cursor_area.x + cursor_area.width;
+	int y1 = parent.y + cursor_area.y;
+	int y2 = parent.y + cursor_area.y + cursor_area.height;
+	int x = x1;
+	int y = y2;
+
+	int available_right = output_box.x + output_box.width - x1;
+	int available_left = x2 - output_box.x;
+	if (available_right < popup_width && available_left > available_right) {
+		x = x2 - popup_width;
+	}
+
+	int available_down = output_box.y + output_box.height - y2;
+	int available_up = y1 - output_box.y;
+	if (available_down < popup_height && available_up > available_down) {
+		y = y1 - popup_height;
+	}
+
+	wlr_scene_node_set_position(&relative->node, x - parent.x - geo.x, y - parent.y - geo.y);
+	if (cursor_rect) {
+		struct wlr_box box = {
+			.x = x1 - x,
+			.y = y1 - y,
+			.width = cursor_area.width,
+			.height = cursor_area.height,
+		};
+		wlr_input_popup_surface_v2_send_text_input_rectangle(
+			popup->popup_surface, &box);
+	}
+	wlr_scene_node_set_position(&popup->scene_tree->node, x - geo.x, y - geo.y);
+}
+
+static void input_popup_set_focus(struct sway_input_popup *popup,
+		struct wlr_surface *surface) {
+	wl_list_remove(&popup->focused_surface_unmap.link);
+
+	if (surface == NULL) {
+		wl_list_init(&popup->focused_surface_unmap.link);
+		input_popup_update(popup);
+		return;
+	}
+	struct wlr_layer_surface_v1 *layer_surface =
+		wlr_layer_surface_v1_try_from_wlr_surface(surface);
+	if (layer_surface != NULL) {
+		wl_signal_add(
+			&layer_surface->surface->events.unmap, &popup->focused_surface_unmap);
+		input_popup_update(popup);
+		return;
+	}
+
+	struct sway_view *view = view_from_wlr_surface(surface);
+	wl_signal_add(&view->events.unmap, &popup->focused_surface_unmap);
+}
+
+static void handle_im_popup_destroy(struct wl_listener *listener, void *data) {
+	struct sway_input_popup *popup =
+		wl_container_of(listener, popup, popup_destroy);
+	wl_list_remove(&popup->focused_surface_unmap.link);
+	wl_list_remove(&popup->popup_surface_commit.link);
+	wl_list_remove(&popup->popup_destroy.link);
+	wl_list_remove(&popup->link);
+	if (popup->scene_tree != NULL) {
+		wlr_scene_node_destroy(&popup->scene_tree->node);
+	}
```

Comment:
I believe destroying `popup->scene_tree` here is unnecessary because we already do this in the `unmap` event handler via `input_popup_update()`.